model/map-jpa/src/main/java/org/keycloak/models/map/jpa/client/storage/JpaClientMapStorageProviderFactory.java

```diff
@@ -0,0 +1,224 @@
+
+    private volatile EntityManagerFactory emf;
+
+    private static final Logger logger = Logger.getLogger(JpaClientMapStorageProviderFactory.class);
+
+    private Config.Scope config;
+
+    @Override
+    public MapStorageProvider create(KeycloakSession session) {
+        lazyInit(session);
+
+        return new JpaClientMapStorageProvider(emf.createEntityManager());
+    }
+
+    @Override
+    public void init(Config.Scope config) {
+        this.config = config;
+    }
+
+    @Override
+    public void postInit(KeycloakSessionFactory factory) {
+    }
+
+    @Override
+    public String getId() {
+        return PROVIDER_ID;
+    }
+
+    @Override
+    public String getHelpText() {
+        return "Initial atempt to create JpaMapStorage: TODO";
+    }
+
+    @Override
+    public boolean isSupported() {
+        return Profile.isFeatureEnabled(Profile.Feature.MAP_STORAGE);
+    }
+
+    @Override
+    public void close() {
+        if (emf != null) {
+            emf.close();
+        }
+    }
+
+    private void lazyInit(KeycloakSession session) {
+        if (emf == null) {
+            synchronized (this) {
+                if (emf == null) {
+                    logger.debugf("Initializing JPA connections%s", StackUtil.getShortStackTrace());
+
+                    Map<String, Object> properties = new HashMap<>();
+
+                    String unitName = "keycloak-client-store";
+
+                    String dataSource = config.get("dataSource");
+                    if (dataSource != null) {
+                        properties.put(AvailableSettings.JPA_NON_JTA_DATASOURCE, dataSource);
+                    } else {
+                        properties.put(AvailableSettings.JPA_JDBC_URL, config.get("url"));
+                        properties.put(AvailableSettings.JPA_JDBC_DRIVER, config.get("driver"));
+
+                        String user = config.get("user");
+                        if (user != null) {
+                            properties.put(AvailableSettings.JPA_JDBC_USER, user);
+                        }
+                        String password = config.get("password");
+                        if (password != null) {
+                            properties.put(AvailableSettings.JPA_JDBC_PASSWORD, password);
+                        }
+                    }
+
+                    String schema = config.get("schema");
+                    if (schema != null) {
+                        properties.put(JpaUtils.HIBERNATE_DEFAULT_SCHEMA, schema);
+                    }
+
+                    properties.put("hibernate.show_sql", config.getBoolean("showSql", false));
+                    properties.put("hibernate.format_sql", config.getBoolean("formatSql", true));
+                    properties.put("hibernate.dialect", config.get("driverDialect"));
+
+                    properties.put(AvailableSettings.ISOLATION, String.valueOf(Connection.TRANSACTION_REPEATABLE_READ));
```

Comment:
@hmlnarik I wanted to highlight this change, with default isolation level, concurrent tests fails, but I believe we cannot set the level to `TRANSACTION_REPEATABLE_READ`, right?

Reply:
I think that instead of setting it, we should check the transaction level and log a warning if it is lower than that. Perhaps